src/NHibernate.Test/NHSpecificTest/GH2029/Fixture.cs

```diff
@@ -0,0 +1,181 @@
+				});
+
+				tx.Commit();
+			}
+		}
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var tx = session.BeginTransaction())
+			{
+				session.CreateQuery("delete from TestClass").ExecuteUpdate();
+
+				tx.Commit();
+			}
+		}
+
+		[Test]
+		public void NullableIntOverflow()
+		{
+			using (var session = OpenSession())
+			using (session.BeginTransaction())
+			using (var sqlLog = new SqlLogSpy())
+			{
+				var groups = session.Query<TestClass>()
+					   .GroupBy(i => 1)
+					   .Select(g => new
+					   {
+						   s = g.Sum(i => (long) i.NullableInt32Prop)
+					   })
+					   .ToArray();
+
+				Assert.That(FindAllOccurrences(sqlLog.GetWholeLog(), "cast"), Is.EqualTo(1));
+				Assert.That(groups, Has.Length.EqualTo(1));
+				Assert.That(groups[0].s, Is.EqualTo((long) int.MaxValue * 2));
+			}
+		}
+
+		[Test]
+		public void IntOverflow()
+		{
+			using (var session = OpenSession())
+			using (session.BeginTransaction())
+			using (var sqlLog = new SqlLogSpy())
+			{
+				var groups = session.Query<TestClass>()
+									.GroupBy(i => 1)
+									.Select(g => new
+									{
+										s = g.Sum(i => (long) i.Int32Prop)
+									})
+									.ToArray();
+
+				Assert.That(FindAllOccurrences(sqlLog.GetWholeLog(), "cast"), Is.EqualTo(1));
+				Assert.That(groups, Has.Length.EqualTo(1));
+				Assert.That(groups[0].s, Is.EqualTo((long) int.MaxValue * 3));
+			}
+		}
+
+		[Test]
+		public void NullableInt64NoCast()
+		{
+			using (var session = OpenSession())
+			using (session.BeginTransaction())
+			using (var sqlLog = new SqlLogSpy())
+			{
+				var groups = session.Query<TestClass>()
+				                    .GroupBy(i => 1)
+				                    .Select(g => new {
+					                    s = g.Sum(i => i.NullableInt64Prop)
+				                    })
+				                    .ToArray();
+
+				Assert.That(sqlLog.GetWholeLog(), Does.Not.Contains("cast"));
+				Assert.That(groups, Has.Length.EqualTo(1));
+				Assert.That(groups[0].s, Is.EqualTo((long) int.MaxValue * 2));
+			}
+		}
+
+		[Test]
+		public void Int64NoCast()
+		{
+			using (var session = OpenSession())
+			using (session.BeginTransaction())
+			using (var sqlLog = new SqlLogSpy())
+			{
+				var groups = session.Query<TestClass>()
+				                    .GroupBy(i => 1)
+				                    .Select(g => new {
+					                    s = g.Sum(i => i.Int64Prop)
+				                    })
+				                    .ToArray();
+
+				Assert.That(sqlLog.GetWholeLog(), Does.Not.Contains("cast"));
+				Assert.That(groups, Has.Length.EqualTo(1));
+				Assert.That(groups[0].s, Is.EqualTo((long) int.MaxValue * 3));
+			}
+		}
+
+		private int FindAllOccurrences(string source, string substring)
+		{
+			if (source == null)
+			{
+				return 0;
+			}
+			int n = 0, count = 0;
+			while ((n = source.IndexOf(substring, n, StringComparison.InvariantCulture)) != -1)
+			{
+				n += substring.Length;
+				++count;
```

Comment:
In c#, pre-increment is not better than post-increment, performance wise, while being generally considered less readable.
So better use the more usual `count++;` syntax.